File: src/android/app/src/main/kotlin/com/austa/superapp/features/healthrecords/domain/models/HealthRecord.kt
package com.austa.superapp.features.healthrecords.domain.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import com.google.gson.annotations.SerializedName // v2.10.1

/**
 * Enumeration of FHIR-aligned health record types with standardized coding.
 */
enum class HealthRecordType {
    @SerializedName("consultation")
    CONSULTATION,
    
    @SerializedName("lab_result")
    LAB_RESULT,
    
    @SerializedName("prescription")
    PRESCRIPTION,
    
    @SerializedName("imaging")
    IMAGING,
    
    @SerializedName("vital_signs")
    VITAL_SIGNS,
    
    @SerializedName("wearable_data")
    WEARABLE_DATA,
    
    @SerializedName("immunization")
    IMMUNIZATION,
    
    @SerializedName("procedure")
    PROCEDURE,
    
    @SerializedName("condition")
    CONDITION,
    
    @SerializedName("medication")
    MEDICATION
}

/**
 * Enumeration of FHIR-compliant record statuses with audit support.
 */
enum class HealthRecordStatus {
    @SerializedName("draft")
    DRAFT,
    
    @SerializedName("final")
    FINAL,
    
    @SerializedName("amended")
    AMENDED,
    
    @SerializedName("deleted")
    DELETED,
    
    @SerializedName("entered_in_error")
    ENTERED_IN_ERROR,
    
    @SerializedName("preliminary")
    PRELIMINARY
}

/**
 * Enhanced metadata class with comprehensive audit and security tracking.
 */
@Parcelize
data class HealthRecordMetadata(
    @SerializedName("version")
    val version: Int,
    
    @SerializedName("created_at")
    val createdAt: String,
    
    @SerializedName("created_by")
    val createdBy: String,
    
    @SerializedName("updated_at")
    val updatedAt: String,
    
    @SerializedName("updated_by")
    val updatedBy: String,
    
    @SerializedName("facility")
    val facility: String,
    
    @SerializedName("department")
    val department: String,
    
    @SerializedName("access_history")
    val accessHistory: List<AccessEntry>,
    
    @SerializedName("encryption_status")
    val encryptionStatus: Boolean,
    
    @SerializedName("data_retention_policy")
    val dataRetentionPolicy: String,
    
    @SerializedName("compliance_flags")
    val complianceFlags: List<String>
) : Parcelable

/**
 * Secure attachment handling with content validation and encryption.
 */
@Parcelize
data class HealthRecordAttachment(
    @SerializedName("id")
    val id: String,
    
    @SerializedName("type")
    val type: String,
    
    @SerializedName("title")
    val title: String,
    
    @SerializedName("content_type")
    val contentType: String,
    
    @SerializedName("size")
    val size: Long,
    
    @SerializedName("url")
    val url: String,
    
    @SerializedName("uploaded_at")
    val uploadedAt: String,
    
    @SerializedName("uploaded_by")
    val uploadedBy: String,
    
    @SerializedName("checksum")
    val checksum: String,
    
    @SerializedName("encryption_key")
    val encryptionKey: String,
    
    @SerializedName("access_control")
    val accessControl: List<String>,
    
    @SerializedName("retention_period")
    val retentionPeriod: String
) : Parcelable

/**
 * Access entry for tracking record access history.
 */
@Parcelize
data class AccessEntry(
    @SerializedName("timestamp")
    val timestamp: String,
    
    @SerializedName("user_id")
    val userId: String,
    
    @SerializedName("action")
    val action: String,
    
    @SerializedName("ip_address")
    val ipAddress: String
) : Parcelable

/**
 * Core data class representing a FHIR R4 compliant health record with comprehensive PHI protection.
 */
@Parcelize
data class HealthRecord(
    @SerializedName("id")
    val id: String,
    
    @SerializedName("patient_id")
    val patientId: String,
    
    @SerializedName("provider_id")
    val providerId: String,
    
    @SerializedName("type")
    val type: HealthRecordType,
    
    @SerializedName("date")
    val date: String,
    
    @SerializedName("content")
    val content: Map<String, Any>,
    
    @SerializedName("metadata")
    val metadata: HealthRecordMetadata,
    
    @SerializedName("attachments")
    val attachments: List<HealthRecordAttachment>,
    
    @SerializedName("status")
    val status: HealthRecordStatus,
    
    @SerializedName("version")
    val version: Int,
    
    @SerializedName("security_labels")
    val securityLabels: List<String>,
    
    @SerializedName("confidentiality")
    val confidentiality: String,
    
    @SerializedName("signature")
    val signature: String
) : Parcelable {

    /**
     * Converts health record to FHIR-compliant JSON string with security handling.
     * @return FHIR R4 compliant JSON representation
     */
    fun toJson(): String {
        // Implementation would include:
        // - FHIR compliance validation
        // - Security label and confidentiality application
        // - ISO8601 date formatting
        // - Attachment processing and validation
        // - Digital signature generation
        // - JSON string formatting
        throw NotImplementedError("Implementation required")
    }

    companion object {
        /**
         * Creates health record from FHIR-compliant JSON with validation.
         * @param jsonString FHIR R4 compliant JSON string
         * @return Validated health record instance
         */
        fun fromJson(jsonString: String): HealthRecord {
            // Implementation would include:
            // - JSON structure parsing and validation
            // - FHIR compliance verification
            // - Security constraint validation
            // - Secure attachment processing
            // - Digital signature verification
            // - HealthRecord instance creation
            throw NotImplementedError("Implementation required")
        }
    }
}

File: src/web/src/lib/types/healthRecord.ts
/**
 * @fileoverview TypeScript types and interfaces for health record data structures
 * Implements FHIR R4 compliant data models with enhanced PHI protection
 * @version 1.0.0
 */

// External imports - v5.0.0
import { ErrorCode } from '../constants/errorCodes';

/**
 * Comprehensive enum defining all supported types of health records
 */
export enum HealthRecordType {
  CONSULTATION = 'CONSULTATION',
  LAB_RESULT = 'LAB_RESULT',
  PRESCRIPTION = 'PRESCRIPTION',
  IMAGING = 'IMAGING',
  VITAL_SIGNS = 'VITAL_SIGNS',
  WEARABLE_DATA = 'WEARABLE_DATA'
}

/**
 * Enum defining possible health record statuses for lifecycle management
 */
export enum HealthRecordStatus {
  DRAFT = 'DRAFT',
  FINAL = 'FINAL',
  AMENDED = 'AMENDED',
  DELETED = 'DELETED'
}

/**
 * HIPAA compliance tracking interface
 */
interface IHIPAACompliance {
  isProtectedHealth: boolean;
  dataMinimizationApplied: boolean;
  encryptionVerified: boolean;
  accessRestrictions: string[];
  lastComplianceCheck: Date;
  complianceOfficer: string;
}

/**
 * Audit trail entry interface
 */
interface IAuditEntry {
  timestamp: Date;
  action: string;
  performedBy: string;
  ipAddress: string;
  changes?: Record<string, any>;
}

/**
 * Enhanced metadata interface with HIPAA compliance tracking and audit capabilities
 */
export interface IHealthRecordMetadata {
  version: number;
  createdAt: Date;
  createdBy: string;
  updatedAt: Date;
  updatedBy: string;
  facility: string;
  department: string;
  hipaaCompliance: IHIPAACompliance;
  auditTrail: IAuditEntry[];
}

/**
 * Interface for secure health record attachments with encryption and integrity verification
 */
export interface IHealthRecordAttachment {
  id: string;
  type: string;
  title: string;
  contentType: string;
  size: number;
  url: string;
  uploadedAt: Date;
  uploadedBy: string;
  encryptionDetails: {
    algorithm: string;
    keyId: string;
    initVector: string;
  };
  securityHash: string;
}

/**
 * Comprehensive interface for health record data structure with enhanced security features
 */
export interface IHealthRecord {
  id: string;
  patientId: string;
  providerId: string;
  type: HealthRecordType;
  date: Date;
  content: Record<string, any>;
  metadata: IHealthRecordMetadata;
  attachments: IHealthRecordAttachment[];
  status: HealthRecordStatus;
  securityClassification: string;
  encryptionLevel: string;
}

/**
 * FHIR R4 Resource Types supported by the system
 */
export enum FHIRResourceType {
  PATIENT = 'Patient',
  OBSERVATION = 'Observation',
  DIAGNOSTIC_REPORT = 'DiagnosticReport',
  MEDICATION_REQUEST = 'MedicationRequest',
  IMAGING_STUDY = 'ImagingStudy',
  DEVICE_METRIC = 'DeviceMetric'
}

/**
 * Interface for wearable device data integration
 */
export interface IWearableData {
  deviceId: string;
  deviceType: string;
  timestamp: Date;
  metrics: {
    heartRate?: number;
    bloodPressure?: {
      systolic: number;
      diastolic: number;
    };
    bloodOxygen?: number;
    steps?: number;
    temperature?: number;
    sleep?: {
      startTime: Date;
      endTime: Date;
      quality: string;
    };
  };
  calibrationStatus: string;
  accuracy: number;
}

/**
 * Error types specific to health record operations
 */
export enum HealthRecordErrorType {
  INVALID_RECORD = ErrorCode.INVALID_INPUT,
  UNAUTHORIZED_ACCESS = ErrorCode.FORBIDDEN,
  PHI_VALIDATION_ERROR = ErrorCode.HIPAA_VIOLATION
}

/**
 * Security classification levels for health records
 */
export enum SecurityClassification {
  PUBLIC = 'PUBLIC',
  RESTRICTED = 'RESTRICTED',
  CONFIDENTIAL = 'CONFIDENTIAL',
  HIGHLY_CONFIDENTIAL = 'HIGHLY_CONFIDENTIAL'
}

/**
 * Encryption levels for different types of health data
 */
export enum EncryptionLevel {
  STANDARD = 'AES-128',
  ENHANCED = 'AES-256',
  MAXIMUM = 'AES-256-GCM'
}